### DISCUSSION TITLE
Nextflow integration

### DISCUSSION THREAD

**Discussion by igboyes** [virtool/.github#4] [Ideas]
From @peterk87:

[Nextflow](https://www.nextflow.io/) is a workflow execution manager and workflow specification language with a lot of nice features for developing and executing bioinformatics workflows (e.g. support for Conda, containers, different execution environments).

There are a lot of benefits for having the bioinformatics workflows de-coupled from Virtool such as ease of testing, development, workflow versioning and usage/evaluation.

If Virtool could understand how to execute Nextflow workflows, track their progress and how to process their outputs, then a plethora of well established workflows (e.g. [nf-core](https://nf-co.re/)) could be used from Virtool to perform a wide variety of bioinformatics analyses.

**NB:** *This is a bit of a brain dump for how to integrate Nextflow with Virtool and use it for generic bioinformatics workflow execution*

# Challenges

- Workflow input parameters specification and validation
  - Require `nextflow_schema.json` JSON schema file? Require `--input` parameter with standard input type?
- Workflow output file specification for processing and visualization in Virtool
  - JSON schema for output results directory structure and file types present?
- Analysis output file provenance information
  - Ensure that every output file can be linked back to a specific task within the workflow.
- Task resource allocation
  - each workflow process should have its required resources specified and it should be possible to customize these with a custom Nextflow config handled by Virtool
- Workflow task retry
  - most Nextflow workflows should have built-in config details on retrying tasks in case of errors, however, dependent on the execution environment (e.g. Slurm) some error codes may not be captured so retrying based on those codes may not be possible
- HPC cluster/cloud execution
  - e.g. workflow execution on an HPC cluster with the Slurm scheduler, or cloud using K8S

# Nextflow web log API

Nextflow has support for [web log via HTTP](https://www.nextflow.io/docs/latest/tracing.html#weblog-via-http) to a server API endpoint for logging and tracking the status of executed Nextflow workflows.

[Nextflow Tower](https://tower.nf) is an example of how the web log API can be used.

# Nextflow & Virtool

Below is a rough series of steps that Virtool could go through to prepare, trigger, monitor and process a Nextflow workflow execution:

- user selects sample(s) and Virtool compatible Nextflow workflow and version of workflow
- pull latest/correct version of Virtool compatible Nextflow workflow
- retrieve Nextflow workflow `nextflow_schema.json`
- dynamically render workflow input params UI in Virtool using `nextflow_schema.json`
- have user enter workflow params, validate input, have Virtool automatically fill in what it can using user sample collection and data available as well as based on what kind of input workflow accepts
- stage workflow execution directory symlinking input data as appropriate
- Start Nextflow workflow execution with
  - URL to Virtool HTTP API endpoint
  - unique workflow execution name (random UUID?) for distinguishing between different analyses
  - workflow parameters (input files, work dir, results dir, workflow parameters, correct workflow profile(s))
  - custom Nextflow config (executor and params, caching dir, task resource allocations)
- capture, process and visualize Nextflow workflow execution progress
- **On completion**
  - process output files in published results directory for information that will be saved into the Virtool DB
  - generate link to zip archive of workflow execution results directory
  - render analysis outputs as appropriate in Virtool (e.g. VCF/SnpEff output to genome diagram with variants visualized as well as interactive table; phylogenetic tree viz with variant and sample metadata highlighting (e.g. [Nextstrain](https://nextstrain.org/ncov/global)); Kraken2 classification results visualized with Krona and/or Sankey diagram)
  - clean-up work directory while ensuring symlinks to work dir files/dirs in results directory are copied (Nextflow workflows should specify a `publish_dir_mode` param that can be set to `copy` so that all files are copied to the results output dir)
- **On error**
  - show verbose error information including if possible task command line, stderr/stdout, error code (what that error code may mean)
  - maybe allow user to download intermediate output files so that they can troubleshoot and find out why their analysis failed

**Comment by MikeERott** [comment]
This is interesting. Ian and I had discussed the idea of being able to easily bring in alternative workflows into Virtool. Sounds like this could be a solution. I would have to understand it better as it looks like it looks complicated and a major project  feature to implement. If it were to be implemented, how easy would it be for someone to build a nextflow workflow and bring it into Virtool?
